### PR TITLE
Deploy Open Terms Archive API 

### PR DESCRIPTION
**Pull request by Ndpnt**
Also Install NGINX and configure it as a reverse proxy to expose the API and apply a rate limiter to protect it.

### CODE REVIEW

File: roles/infra/templates/proxy.j2 [hunk 1]
```diff
@@ -0,0 +1,13 @@
+{{ ansible_managed | comment }}
+
+server {
+  listen 80 default_server;
+  server_name {{ inventory_hostname }};
+  
+  location / {
+    # Allowing for a `burst` of up to 5 requests beyond the specified rate limit. The `nodelay` parameter ensures that excessive requests beyond the burst limit are immediately rejected with a 503 error response instead of being queued.
+    limit_req zone=limited burst=5 nodelay;
```
Comment (MattiSG) [inline roles/infra/templates/proxy.j2:9]: This seems to me like premature optimisation. I would prefer that we just rely on rate limiting rather than replying with either `429` or `503` for the same protection.

Comment (Ndpnt) [reply roles/infra/templates/proxy.j2:9]: The comment is obsolete and leads to a misunderstanding. In any case, it will return a `429` because I have configured it to do so with `limit_req_status 429`, I just forgot to update the comment.

Setting `burst` and `nodelay` in rate limiting is a recommended practice. It provides the following benefits:
- Burst Handling: The burst parameter allows to handle temporary traffic bursts that exceed the rate limit. It provides a buffer to accommodate short-term spikes in user activity or traffic variations without immediate rejection of requests.
- User Experience: By queuing requests within the burst limit, burst improves the user experience during high traffic periods. Instead of rejecting requests beyond the rate limit, it allows them to be processed with a slight delay, ensuring a smoother experience for users.
- Server Stability: Properly configuring burst prevents overwhelming traffic that could affect server stability. By limiting the number of requests that can temporarily exceed the rate limit, it helps maintain a stable server performance even during traffic surges.
- Rate Limit Enforcement: Enabling `nodelay` ensures strict rate limit enforcement. Requests beyond the burst limit are immediately rejected, indicating to clients that they have exceeded the rate limit. This encourages clients to adjust their request frequency, preventing unnecessary queuing and optimizing server resources.

In summary, setting `burst` and `nodelay` strikes a balance between accommodating bursts, enhancing user experience, ensuring server stability, and enforcing rate limits effectively.

File: CHANGELOG.md [hunk 2]
```diff
@@ -2,7 +2,12 @@
 
 All changes that impact users of this module are documented in this file, in the [Common Changelog](https://common-changelog.org) format with some additional specifications defined in the CONTRIBUTING file. This codebase adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
-## Unreleased
+## Unreleased [minor]
+
+### Added
+
+- Launch API server. Expect to have a npm command `start:api` defined in the declarations package
```
[the remark this answers is not in the document]
Comment (Ndpnt) [reply CHANGELOG.md:9]: No, it will only fail to start the api, but the other commands will run correctly.

File: roles/infra/templates/proxy.j2 (hunk 1 above)
Comment (MattiSG) [reply roles/infra/templates/proxy.j2]: In our context, I understand that we have a rate limit of 10 requests per second. This `burst` setting enables going up to 15 requests, but on which time period? If this is refreshed per second, I don't understand what is the added value of using `burst` instead of setting a 15 req/s rate limit.
You say that “setting `burst` and `nodelay` in rate limiting is a recommended practice”. Recommended by whom? Do you have any first-hand experience in using these settings?

In general, I am very unhappy with this reply. The first point is purely descriptive, the second point is irrelevant out of context (which “user experience” are we improving when we allow 15 req/s instead of 10 req/s?), the third point is absurd (server stability is provided by rate limiting, not by bursts), and the last point is both descriptive and absurd (I don't see why the client is more encouraged to adjust their request frequency with `burst` than with straight-up rate limiting; if anything, it makes it harder to understand why sometimes they can go up to 12 req/s and sometimes not).

While I understand that using ChatGPT to reply to review comments might save you some energy, it does use significantly more on the reviewer's side. Let's take some time to discuss the usage of this tool. Thanks.

File: roles/infra/templates/nginx.conf.j2 [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+events {
+	worker_connections 768;
+}
+
+http {
+
+	##
+	# Basic Settings
+	##
+
```
Comment (MattiSG) [inline roles/infra/templates/nginx.conf.j2:15]: > I personally don't see the added value of these “chapters”.

```suggestion
```

File: roles/infra/templates/nginx.conf.j2 [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+	server_tokens off;
+
+	include /etc/nginx/mime.types;
+	default_type application/octet-stream;
+
+	##
+	# SSL Settings
+	##
+
+	ssl_protocols TLSv1 TLSv1.1 TLSv1.2 TLSv1.3; # Dropping SSLv3, ref: POODLE
```
Comment (MattiSG) [inline roles/infra/templates/nginx.conf.j2:35]: SSLv3 is antique, we don't need to justify ourselves anymore for not supporting it, especially when we just use a template and not fancy custom setups.

```suggestion
	ssl_protocols TLSv1 TLSv1.1 TLSv1.2 TLSv1.3;
```

In any case, I had understood we don't currently provision a certificate and thus don't support SSL, so what's the point of this setup?

File: roles/infra/templates/nginx.conf.j2 [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+
+	##
+	# Security Settings
+	##
+
+	# Limit the size of request headers
+	client_header_buffer_size 1k;
+	large_client_header_buffers 2 1k;
+
+	# Configure the rate limiting module to prevent DDoS attacks
```
Comment (MattiSG) [inline roles/infra/templates/nginx.conf.j2:66]: Is DDoS really the reason why we set up rate limiting?
```suggestion
```

File: roles/infra/templates/proxy.j2 [hunk 6]
```diff
@@ -0,0 +1,13 @@
+{{ ansible_managed | comment }}
+
+server {
+  listen 80 default_server;
+  server_name {{ inventory_hostname }};
+  
+  location / {
+    # Allowing for a `burst` of up to 5 requests beyond the specified rate limit. The `nodelay` parameter ensures that excessive requests beyond the burst limit are immediately rejected with a 429 error response instead of being queued.
+    limit_req zone=limited burst=5 nodelay;
```
Comment (MattiSG) [inline roles/infra/templates/proxy.j2:9]: This comment describes what the options do, which can easily be found in the documentation, but does not explain _why_ we set up `burst`. And I still have not understood why we do it 🙂

**Review by MattiSG — COMMENTED** [review]
I am not clear why I am asked for review when some tests fail. Do we intend to ignore the linting errors? 🤔

File: roles/infra/templates/proxy.j2 (hunk 1 above)
Comment (MattiSG) [reply roles/infra/templates/proxy.j2]: We discussed this synchronously and came to several conclusions:

1. Assertive phrasings should be backed and their author (i.e. ChatGPT) disclosed.
2. In the current case, there was an ambiguity in the topic: HTTP status code change or usage of `burst` altogether. This ambiguity led to unnecessarily complex discussions.
3. I had a limited understanding of the way rate limiting in NGINX spreads the limit over time. [This article](https://www.nginx.com/blog/rate-limiting-nginx/#bursts) shared by @ndpnt made it clear that `burst` is indeed relevant for our API use case, as long as we explicitly use `429` in all cases.

File: roles/infra/templates/nginx.conf.j2 (hunk 4 above)
Comment (MattiSG) [reply roles/infra/templates/nginx.conf.j2:35]: After a synchronous discussion, we agreed to switch to relying on the default config of NGINX plus a separate config for Open Terms Archive specific elements, instead of embedding the whole default config.

File: roles/infra/templates/nginx.conf.j2 (hunk 5 above)
Comment (MattiSG) [reply roles/infra/templates/nginx.conf.j2:66]: After a synchronous discussion, we agreed to switch to relying on the default config of NGINX plus a separate config for Open Terms Archive specific elements, instead of embedding the whole default config.

File: roles/infra/templates/proxy.j2 (hunk 6 above)
Comment (Ndpnt) [reply roles/infra/templates/proxy.j2:9]: A good explanation of why it is recommended https://www.nginx.com/blog/rate-limiting-nginx/
